File: healthcare/api/nurse_task.py
import json
import frappe
from frappe import _
from frappe.utils import cint, now_datetime


ADMIN_ROLES = {'Administrator', 'System Manager', 'Healthcare Administrator', 'Website Manager'}


def _is_admin() -> bool:
	return bool(ADMIN_ROLES & set(frappe.get_roles(frappe.session.user)))


def _get_practitioner_for_user(user: str) -> str | None:
	"""Return the Healthcare Practitioner name linked to the given user account."""
	return frappe.db.get_value("Healthcare Practitioner", {"user": user}, "name")


@frappe.whitelist()
def create_nurse_task(
	patient: str,
	task_type: str,
	scheduled_time: str,
	description: str = "",
	priority: str = "Routine",
	assigned_nurse: str | None = None,
	shift: str | None = None,
	cost_center: str | None = None,
	due_time: str | None = None,
	medication: str | None = None,
	dosage: str | None = None,
	route: str | None = None,
	frequency: str | None = None,
	is_prn: int = 0,
	prn_indication: str | None = None,
	min_interval_hours: float | None = None,
	notes: str | None = None,
	reference_doctype: str | None = None,
	encounter: str | None = None,
) -> dict:
	"""Create a single Nurse Task document."""
	if not patient:
		frappe.throw(_("Patient is required"))
	if not task_type:
		frappe.throw(_("Task Type is required"))
	if not scheduled_time:
		frappe.throw(_("Scheduled Time is required"))

	doc = frappe.new_doc("Nurse Task")
	doc.patient = patient
	doc.task_type = task_type
	doc.scheduled_time = scheduled_time
	doc.description = description or ""
	doc.priority = priority or "Routine"
	doc.status = "Pending"

	if assigned_nurse:
		doc.assigned_nurse = assigned_nurse
	if shift:
		doc.shift = shift
	if cost_center:
		doc.cost_center = cost_center
	if due_time:
		doc.due_time = due_time
	if medication:
		doc.medication = medication
	if dosage:
		doc.dosage = dosage
	if route:
		doc.route = route
	if frequency:
		doc.frequency = frequency

	doc.is_prn = cint(is_prn)
	if prn_indication:
		doc.prn_indication = prn_indication
	if min_interval_hours is not None:
		doc.min_interval_hours = min_interval_hours
	if notes:
		doc.notes = notes
	if reference_doctype:
		doc.reference_doctype = reference_doctype
	if encounter:
		doc.encounter = encounter

	doc.insert(ignore_permissions=True)
	return {"name": doc.name}


@frappe.whitelist()
def bulk_create_nurse_tasks(tasks: str | list) -> dict:
	"""Create multiple Nurse Tasks at once.

	``tasks`` is either a JSON string or a list of dicts, each matching the
	parameters accepted by :func:`create_nurse_task`.
	"""
	if isinstance(tasks, str):
		tasks = json.loads(tasks)

	created = []
	for task_data in tasks:
		result = create_nurse_task(**task_data)
		created.append(result["name"])

	return {"created": created, "count": len(created)}


@frappe.whitelist()
def get_nurse_tasks(
	limit: int = 50,
	offset: int = 0,
	patient: str | None = None,
	status: str | None = None,
	my_tasks: int = 0,
	assigned_nurse: str | None = None,
) -> list:
	"""Return Nurse Task rows.

	Access rules
	------------
	* Admins (System Manager / Healthcare Administrator / Administrator):
	  see every task — no automatic nurse filter is applied.
	* Everyone else (nurses): only see tasks where ``assigned_nurse``
	  matches the Employee record linked to their user account.
	  If their user has no linked Employee record the list is empty.
	"""
	filters: dict = {}

	if patient:
		filters["patient"] = patient
	if status:
		filters["status"] = status

	if not _is_admin():
		# Non-admins always see only their own tasks
		practitioner = _get_practitioner_for_user(frappe.session.user)
		if not practitioner:
			return []
		filters["assigned_nurse"] = practitioner
	else:
		# Admins can optionally be scoped to a specific nurse or their own tasks
		if cint(my_tasks):
			practitioner = _get_practitioner_for_user(frappe.session.user)
			if practitioner:
				filters["assigned_nurse"] = practitioner
		elif assigned_nurse:
			filters["assigned_nurse"] = assigned_nurse

	rows = frappe.get_all(
		"Nurse Task",
		filters=filters,
		fields=[
			"name",
			"patient",
			"task_type",
			"description",
			"priority",
			"status",
			"scheduled_time",
			"due_time",
			"completed_time",
			"assigned_nurse",
			"medication",
			"dosage",
			"route",
			"is_prn",
			"notes",
			"reference_doctype",
			"encounter",
			"completed_by",
			"creation",
		],
		order_by="scheduled_time asc, creation desc",
		limit=cint(limit or 50),
		start=cint(offset or 0),
	)

	# Enrich with patient_name and nurse full name
	for row in rows:
		if row.get("patient"):
			row["patient_name"] = (
				frappe.db.get_value("Patient", row["patient"], "patient_name") or row["patient"]
			)
		if row.get("assigned_nurse"):
			row["assigned_nurse_name"] = (
				frappe.db.get_value("Healthcare Practitioner", row["assigned_nurse"], "practitioner_name")
				or row["assigned_nurse"]
			)
		if row.get("medication"):
			row["medication_name"] = (
				frappe.db.get_value("Item", row["medication"], "item_name") or row["medication"]
			)

	return rows


@frappe.whitelist()
def update_nurse_task_status(name: str, status: str, notes: str | None = None) -> dict:
	"""Update the status of a Nurse Task, and stamp completed fields if done."""
	if not name or not status:
		frappe.throw(_("Task name and status are required"))

	allowed = {"Pending", "In Progress", "Completed", "Missed", "Cancelled"}
	if status not in allowed:
		frappe.throw(_("Invalid status. Allowed: {0}").format(", ".join(allowed)))

	doc = frappe.get_doc("Nurse Task", name)
	doc.status = status

	if status == "Completed":
		doc.completed_time = now_datetime()
		doc.completed_by = frappe.session.user

	if notes is not None:
		doc.notes = notes

	doc.save()
	return {"name": doc.name, "status": doc.status}

File: frontend/src/services/nurseTask.ts
import { apiRequest } from './apiClient'

export interface NurseTask {
  name: string
  patient: string
  patient_name?: string
  task_type: string
  description?: string
  priority?: string
  status: string
  scheduled_time: string
  due_time?: string
  completed_time?: string
  assigned_nurse?: string
  assigned_nurse_name?: string
  shift?: string
  shift_label?: string
  medication?: string
  medication_name?: string
  dosage?: string
  route?: string
  is_prn?: 0 | 1
  prn_indication?: string
  notes?: string
  reference_doctype?: string
  encounter?: string
  completed_by?: string
  creation?: string
}

export interface CreateNurseTaskData {
  patient: string
  task_type: string
  scheduled_time: string
  description?: string
  priority?: string
  /** Healthcare Practitioner name */
  assigned_nurse?: string
  /** Nurse Shift name */
  shift?: string
  /** Branch name */
  cost_center?: string
  due_time?: string
  medication?: string
  dosage?: string
  /** Route of Administration name */
  route?: string
  /** Prescription Frequency name */
  frequency?: string
  is_prn?: boolean
  prn_indication?: string
  min_interval_hours?: number
  notes?: string
  reference_doctype?: string
  encounter?: string
  medication_type?: string
}

export async function createNurseTask(data: CreateNurseTaskData): Promise<{ name: string }> {
  return apiRequest<{ name: string }>(
    '/api/method/healthcare.api.nurse_task.create_nurse_task',
    {
      method: 'POST',
      body: JSON.stringify(data),
    }
  )
}

export async function bulkCreateNurseTasks(tasks: CreateNurseTaskData[]): Promise<{ created: string[]; count: number }> {
  return apiRequest<{ created: string[]; count: number }>(
    '/api/method/healthcare.api.nurse_task.bulk_create_nurse_tasks',
    {
      method: 'POST',
      body: JSON.stringify({ tasks: JSON.stringify(tasks) }),
    }
  )
}

export interface NurseShiftInfo {
  name: string
  label: string
  from_time?: string
  to_time?: string
}

export interface CurrentNurseShiftInfo {
  shift: NurseShiftInfo | null
  window_start?: string | null
  window_end?: string | null
}

export async function fetchCurrentNurseShift(): Promise<CurrentNurseShiftInfo> {
  const res = await fetch('/api/method/healthcare.api.nurse_shift.get_current_nurse_shift_info')
  const data = await res.json()
  if (data?.exc || !res.ok) {
    throw new Error(data?.message || 'Failed to load current nurse shift')
  }
  return (data?.message || { shift: null }) as CurrentNurseShiftInfo
}

export async function fetchNurseShifts(search?: string): Promise<NurseShiftInfo[]> {
  let url = '/api/method/healthcare.api.nurse_shift.get_nurse_shifts'
  if (search) url += `?search=${encodeURIComponent(search)}`
  const res = await fetch(url)
  const data = await res.json()
  if (data?.exc || !res.ok) {
    throw new Error(data?.message || 'Failed to load nurse shifts')
  }
  return Array.isArray(data?.message) ? (data.message as NurseShiftInfo[]) : []
}

export async function fetchNurseTasks(
  options?: {
    limit?: number
    offset?: number
    patient?: string
    status?: string
    task_type?: string
    date_from?: string
    date_to?: string
    my_tasks?: boolean
    assigned_nurse?: string
    shift?: string
    current_shift_only?: boolean
  }
): Promise<NurseTask[]> {
  const params = new URLSearchParams()
  if (options?.limit != null) params.append('limit', String(options.limit))
  if (options?.offset != null) params.append('offset', String(options.offset))
  if (options?.patient) params.append('patient', options.patient)
  if (options?.status) params.append('status', options.status)
  if (options?.task_type) params.append('task_type', options.task_type)
  if (options?.date_from) params.append('date_from', options.date_from)
  if (options?.date_to) params.append('date_to', options.date_to)
  if (options?.my_tasks) params.append('my_tasks', '1')
  if (options?.assigned_nurse) params.append('assigned_nurse', options.assigned_nurse)
  if (options?.shift) params.append('shift', options.shift)
  if (options?.current_shift_only) params.append('current_shift_only', '1')

  const res = await fetch(
    `/api/method/healthcare.api.nurse_task.get_nurse_tasks?${params.toString()}`
  )
  const data = await res.json()

  if (data?.exc || !res.ok) {
    throw new Error(data?.exc || data?.message || 'Failed to load nurse tasks')
  }

  return Array.isArray(data?.message) ? (data.message as NurseTask[]) : []
}

export async function updateNurseTaskStatus(
  name: string,
  status: string,
  notes?: string
): Promise<{ name: string; status: string }> {
  return apiRequest<{ name: string; status: string }>(
    '/api/method/healthcare.api.nurse_task.update_nurse_task_status',
    {
      method: 'POST',
      body: JSON.stringify({ name, status, notes }),
    }
  )
}
